protocols/mdns/src/behaviour/socket.rs

```diff
@@ -0,0 +1,149 @@
+
+    /// Attempts to send data on the socket to a given address.
+    fn poll_send_packet(
+        &mut self,
+        _cx: &mut Context,
+        _packet: &[u8],
+        _to: SocketAddr,
+    ) -> Poll<Result<(), Error>>;
+}
+
+#[cfg(feature = "async-io")]
+pub mod asio {
+    use super::*;
+    use async_io::Async;
+    use futures::FutureExt;
+
+    /// AsyncIo UdpSocket
+    pub type AsyncUdpSocket = Async<UdpSocket>;
+
+    impl AsyncSocket for AsyncUdpSocket {
+        /// Create the async socket from the ```std::net::UdpSocket```
+        fn from_socket(socket: UdpSocket) -> std::io::Result<Self> {
+            Async::new(socket)
+        }
+
+        /// Attempts to receive a single packet on the socket from the remote address to which it is connected.
+        fn poll_receive_packet(
+            &mut self,
+            cx: &mut Context,
+            buf: &mut [u8],
+        ) -> Poll<Result<Option<(usize, SocketAddr)>, Error>> {
+            // Poll receive socket.
+            let _ = futures::ready!(self.poll_readable(cx));
+            match self.recv_from(buf).now_or_never() {
+                Some(Ok((len, from))) => Poll::Ready(Ok(Some((len, from)))),
+                Some(Err(err)) => Poll::Ready(Err(err)),
+                None => Poll::Ready(Ok(None)),
+            }
+        }
+
+        /// Attempts to send data on the socket to a given address.
+        fn poll_send_packet(
+            &mut self,
+            cx: &mut Context,
+            packet: &[u8],
+            to: SocketAddr,
+        ) -> Poll<Result<(), Error>> {
+            let _ = futures::ready!(self.poll_writable(cx));
+            match self.send_to(packet, to).now_or_never() {
+                Some(Ok(_)) => Poll::Ready(Ok(())),
+                Some(Err(err)) => Poll::Ready(Err(err)),
+                None => Poll::Pending,
+            }
+        }
+    }
+}
+
+#[cfg(feature = "tokio")]
+pub mod tokio {
+    use super::*;
+    use ::tokio::net::UdpSocket as TkUdpSocket;
+
+    /// Tokio ASync Socket`
+    pub type TokioUdpSocket = TkUdpSocket;
+
+    impl AsyncSocket for TokioUdpSocket {
+        /// Create the async socket from the ```std::net::UdpSocket```
+        fn from_socket(socket: UdpSocket) -> std::io::Result<Self> {
+            socket.set_nonblocking(true)?;
+            TokioUdpSocket::from_std(socket)
+        }
+
+        /// Attempts to receive a single packet on the socket from the remote address to which it is connected.
+        fn poll_receive_packet(
+            &mut self,
+            cx: &mut Context,
+            buf: &mut [u8],
+        ) -> Poll<Result<Option<(usize, SocketAddr)>, Error>> {
+            match self.poll_recv_ready(cx) {
+                Poll::Ready(Ok(_)) => match self.try_recv_from(buf) {
+                    Ok((len, from)) => Poll::Ready(Ok(Some((len, from)))),
+                    Err(err) if err.kind() == std::io::ErrorKind::WouldBlock => {
+                        Poll::Ready(Ok(None))
```

Comment:
The [docs](https://docs.rs/tokio/1.20.0/tokio/net/struct.UdpSocket.html#errors-2) say this can never happen so I think we could `panic!` here? That should allow us to remove the `Option` type from the return value if I am not mistaken.

Reply:
In Tokio case, from documentation.

[Tokio try_recv_from](https://docs.rs/tokio/1.20.0/tokio/net/struct.UdpSocket.html#method.try_recv_from)

```
// Try to recv data, this may still fail with `WouldBlock`
// if the readiness event is a false positive.
```

Throw panic when is WouldBlock, does not seem to be a good option in this case.

Reply:
Ah you are right, I was mis-reading the code!

What we can then do is put all of this in a loop I think. In case we encounter `WouldBlock`, we can loop up to `poll_recv_ready` which should then return `Pending` and register a Waker. This should allow us to get rid of the `Option` return value!

Thoughts?

Reply:
The other alternative is to also just forward the error to our user. This has likely the advantage that we only have to write the loop once.

I think this is better because even with the loop here, we can't remove the `WouldBlock` case from the signature so any caller may still think it has to handle it.